src/lib/metrics/regression.ts

```diff
@@ -116,20 +106,67 @@ export function mean_squared_error(
     sample_weight: null,
   },
 ): number {
-  const yTrueShape = inferShape(y_true);
-  const yPredShape = inferShape(y_pred);
+  const [yTrueTensor, yPredTensor] = compareShapesAndConvert(y_true, y_pred);
 
-  // Validation 1: empty array check
-  if (yTrueShape[0] === 0 || yPredShape[0] === 0) {
-    throw new TypeError(`y_true ${JSON.stringify(y_true)} and y_pred ${JSON.stringify(y_pred)} cannot be empty`);
+  return tf.losses.meanSquaredError(yTrueTensor, yPredTensor, sample_weight).dataSync()[0];
+}
+
+/**
+ * Mean squared error regression loss
+ *
+ * @example
+ * import { mean_squared_log_error } from 'machinelearn/metrics';
+ *
+ * const y_true = [3, 0.5, 2, 7];
+ * const y_pred = [2.5, 0.0, 2, 8];
+ *
+ * console.log(mean_squared_error(y_true, y_pred));
+ * // result: 0.04902636259794235
+ *
+ * const y_true1 = [[0.5, 1], [1, 1], [7, 6]];
+ * const y_pred1 = [[0, 2], [1, 2], [8, 5]];
+ *
+ * console.log(mean_squared_error(y_true1, y_pred1));
+ * // result: 0.08847352117300034
+ *
+ * @param y_true - Ground truth (correct) target values(should be positive).
+ * @param y_pred - Estimated target values(should be positive).
+ */
+export function mean_squared_log_error(
+  y_true: Type1DMatrix<number> | Type2DMatrix<number> = null,
+  y_pred: Type1DMatrix<number> | Type2DMatrix<number> = null,
+  // Options
+  {
+    /**
+     * Sample weights.
+     */
+    sample_weight = null,
+  }: {
+    sample_weight: number;
+  } = {
+    sample_weight: null,
+  },
+): number {
+  const [yTrueTensor, yPredTensor] = compareShapesAndConvert(y_true, y_pred);
+
+  const error = (y) => new TypeError(`None of the values of ${JSON.stringify(y)} can be less than 0`);
+  if (
+    yTrueTensor
+      .flatten()
+      .arraySync()
+      .filter((a) => a < 0).length > 0
+  ) {
+    throw error(y_true);
   }
 
-  // Validation 2: Same shape
-  if (!isEqual(yTrueShape, yPredShape)) {
-    throw new TypeError(
-      `Shapes of y_true ${JSON.stringify(yTrueShape)} and y_pred ${JSON.stringify(yPredShape)} should be equal`,
-    );
+  if (
+    yPredTensor
+      .flatten()
+      .arraySync()
+      .filter((a) => a < 0).length > 0
```

Comment:
Can we use https://js.tensorflow.org/api/latest/#tf.data.Dataset.filter then inter the shape of the tensor instead of `arraySync` the tensor and using the native array filter?

Reply:
yep, I had my doubts, this actually is better, thanks.

Reply:
ok looked at this option, unfortunately the dimentionality of data supported in tf.data.array is number|number[] and we can expand further, so this might not be possible

Reply:
I realised something, let's not even do this validation at the tensor level, just do it using the raw JS arrays `y_true ` and `y_pred`.

What's the point of converting the tensor back to the raw JS array after turning it into tensors while the arrays are still accessible in the context?

Reply:
Done